tests/Integration/Behaviour/Features/Context/Domain/CartRule/CartRuleFeatureContext.php

```diff
@@ -0,0 +1,305 @@
+     * @throws PrestaShopDatabaseException
+     * @throws PrestaShopException
+     */
+    public function createCartRuleWithReference(string $cartRuleReference, TableNode $node): void
+    {
+        $data = $this->localizeByRows($node);
+        $defaultLanguageId = Configuration::get('PS_LANG_DEFAULT');
+
+        $cartRuleAction = $this->createCartRuleAction(
+            PrimitiveUtils::castStringBooleanIntoBoolean($data['free_shipping']),
+            $data['reduction_percentage'] ?? null,
+            isset($data['reduction_apply_to_discounted_products']) ? PrimitiveUtils::castStringBooleanIntoBoolean($data['reduction_apply_to_discounted_products']) : null,
+            $data['reduction_amount'] ?? null,
+            isset($data['reduction_currency']) ? $this->getSharedStorage()->get($data['reduction_currency']) : null,
+            $data['reduction_tax'] ?? null,
+            $data['gift_product_id'] ?? null,
+            $data['gift_product_attribute_id'] ?? null
+        );
+
+        $command = new AddCartRuleCommand(
+            $data['name'],
+            PrimitiveUtils::castStringBooleanIntoBoolean($data['highlight']),
+            PrimitiveUtils::castStringBooleanIntoBoolean($data['allow_partial_use']),
+            (int) $data['priority'],
+            PrimitiveUtils::castStringBooleanIntoBoolean($data['is_active']),
+            new DateTime($data['valid_from']),
+            new DateTime($data['valid_to']),
+            $data['total_quantity'],
+            $data['quantity_per_user'],
+            $cartRuleAction
+        );
+
+        if (!empty($data['minimum_amount'])) {
+            $currencyId = SharedStorage::getStorage()->get($data['minimum_amount_currency']);
+            $command->setMinimumAmount(
+                $data['minimum_amount'],
+                $currencyId,
+                PrimitiveUtils::castStringBooleanIntoBoolean($data['minimum_amount_tax_included']),
+                PrimitiveUtils::castStringBooleanIntoBoolean($data['minimum_amount_shipping_included'])
+            );
+        }
+
+        $command->setDescription($data['description'] ?? '');
+        $command->setCode($data['code'] ?? '');
+
+        if (isset($data['discount_application_type'])) {
+            $command->setDiscountApplication(
+                $data['discount_application_type'],
+                // if specific product type is provided and product is not, then command should throw exception
+                isset($data['discount_product']) ? $this->getSharedStorage()->get($data['discount_product']) : null
+            );
+        }
+
+        /** @var CartRuleId $cartRuleId */
+        $cartRuleId = $this->getCommandBus()->handle($command);
+        SharedStorage::getStorage()->set($cartRuleReference, $cartRuleId->getValue());
+
+        if (!empty($data['code'])) {
+            // set cart rule id by code when it is not empty
+            SharedStorage::getStorage()->set($data['code'], $cartRuleId->getValue());
```

Comment:
Ah I had almost the same idea (detailed in another comment), I only thought we could add a prefix `cart_rule_code_` for example, it's not mandatory but it could prevent some conflicts maybe

Reply:
Well you'll be aware of it, but a future developer probably not But maybe it's overengineering, let's go with this current colution